Python/ceval.c

```diff
@@ -3417,6 +3417,20 @@ _PyEval_LoadName(PyThreadState *tstate, _PyInterpreterFrame *frame, PyObject *na
     return value;
 }
 
+_PyStackRef
+_PyForIter_NextWithIndex(PyObject *seq, _PyStackRef index)
+{
+    assert(PyStackRef_IsTaggedInt(index));
+    assert(Py_TYPE(seq) == &PyTuple_Type || Py_TYPE(seq) == &PyList_Type);
+    size_t size = Py_SIZE(seq);
+    intptr_t i = PyStackRef_UntagInt(index);
+    if ((size_t)i >= size) {
+        return PyStackRef_NULL;
+    }
+    PyObject *next_o = PySequence_Fast_GET_ITEM(seq, i);
```

Comment:
This isn't thread safe. Check out how to do this in the function `_PyList_GetItemRefNoLock`.